bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceUse.java

```diff
@@ -267,21 +267,21 @@ ServiceUseLock lock() {
 				}
 			}
 		} finally {
-			if (clearAwaitingLock) {
-				AWAITED_LOCKS.remove(Thread.currentThread());
+			if (awaitingThread != null) {
+				AWAITED_LOCKS.remove(awaitingThread);
 			}
 			if (interrupted) {
 				Thread.currentThread().interrupt();
 			}
 		}
 	}
 
-	private static boolean isDeadLocked(ServiceUseLock lock) {
+	private static boolean isDeadLocked(ServiceUseLock lock, Thread currentThread) {
```

Comment:
I would make this `void checkDeadLocked` and have it throw any exception. Then all the deadlock code is together.